### PR TITLE
support sharding parallel

### PR DESCRIPTION
**Pull request by pangengzheng**
### PR types
Others

### PR changes
Others

### Description
Others
card-72028

### CODE REVIEW

**Comment by pangengzheng** [comment]
cherry-pick from https://github.com/PaddlePaddle/Paddle/pull/54069

File: python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer.py [hunk 1]
```diff
@@ -132,14 +128,35 @@ def _map_param_to_rank(self):
...
+                    g_var = param.main_grad
+                if g_var is not None:
+                    g_var.scale_(1.0 / sharding_nrank)
+                    param_rank = self._param2rank[param.name]
+                    paddle.distributed.all_reduce(
+                        g_var,
+                        group=hcg.get_sharding_parallel_group(),
+                        sync_op=True,
+                    )
+                    # TODO(pangengzheng): change to reduce operation when there is no diff in calculating global norm values in HybridParallelClipGrad compared to dp.
```
Comment (FeixLiu) [inline python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer.py:153]: There is a logic conflict between sharding stage 1 and sharding_pp_overlap which can be addressed in one of this two ways:
1. If use `all_reduce`, this part https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/distributed/fleet/meta_parallel/pp_utils/utils.py#L249 should also be updated, you can ref this pr https://github.com/PaddlePaddle/Paddle/pull/54872
2. If use `reduce`, then nothing else this comm method should be updated.

**Review by FeixLiu — CHANGES_REQUESTED** [review]

File: python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer.py (hunk 1 above)
Comment (pangengzheng) [reply python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer.py:153]: ok

**Review by FeixLiu — DISMISSED** [review]
LGTM for the change in pp_utils

**Review by dingjiaweiww — APPROVED** [review]

**Review by sneaxiy — APPROVED** [review]

**Review by qingqing01 — APPROVED** [review]

**Pull request merged**